[jit] Delete the ErrorReport default constructor

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #32880 [jit] de-optionalize SourceRange context
* **#32879 [jit] Delete the ErrorReport default constructor**

An error report without a SourceRange context is bad, because it doesn't
tell the user where something happened. Delete the default constructor
to make it harder to create errors like this (you can still use a fake
SourceRange if you absolutely need to).

Also clean up the only case where the default constructor was used.

Differential Revision: [D19670924](https://our.internmc.facebook.com/intern/diff/D19670924)

suo committed:
[jit] Delete the ErrorReport default constructor

An error report without a SourceRange context is bad, because it doesn't
tell the user where something happened. Delete the default constructor
to make it harder to create errors like this (you can still use a fake
SourceRange if you absolutely need to).

Also clean up the only case where the default constructor was used.

[ghstack-poisoned]

suo committed:
Update on "[jit] Delete the ErrorReport default constructor"


An error report without a SourceRange context is bad, because it doesn't
tell the user where something happened. Delete the default constructor
to make it harder to create errors like this (you can still use a fake
SourceRange if you absolutely need to).

Also clean up the only case where the default constructor was used.

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 1e32fccb:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=1e32fccb5e06c13fecc5aabc4a7ab2e5db56ef59).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4392692) pytorch_linux_xenial_py3_6_gcc5_4_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=93903518) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=93903518))
<details>
<summary>
<code>Automatic merge failed; fix conflicts and then commit the result.</code>
</summary>

```
do so (now or later) by using -b with the checkout command again. Example: 
 
  git checkout -b <new-branch-name> 
 
HEAD is now at 1e32fccb5e Update on "[jit] Delete the ErrorReport default constructor" 
+ git reset --hard 1e32fccb5e06c13fecc5aabc4a7ab2e5db56ef59 
HEAD is now at 1e32fccb5e Update on "[jit] Delete the ErrorReport default constructor" 
+ git merge --allow-unrelated-histories --no-edit --no-ff 7b65acdf9e8345d2ffbbc3b85ae6667f7260407e 
Auto-merging torch/csrc/jit/script/error_report.cpp 
CONFLICT (content): Merge conflict in torch/csrc/jit/script/error_report.cpp 
Automatic merge failed; fix conflicts and then commit the result. 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

suo committed:
Update on "[jit] Delete the ErrorReport default constructor"


An error report without a SourceRange context is bad, because it doesn't
tell the user where something happened. Delete the default constructor
to make it harder to create errors like this (you can still use a fake
SourceRange if you absolutely need to).

Also clean up the only case where the default constructor was used.

Differential Revision: [D19670924](https://our.internmc.facebook.com/intern/diff/D19670924)

[ghstack-poisoned]

eellison commented on torch/csrc/jit/function.cpp:
```diff
@@ -46,19 +45,11 @@ IValue Function::operator()(
-                                // debugging
-        << " method '" << name() << "' is called recursively. "
-        << "Recursive calls are not supported";
+  if (function_creator_) {
```
What happened to `re-establish callsite info for debugging` ?

suo replied:
The callsite info is re-established by passing `loc` to `ErrorReport`.

eellison commented on torch/csrc/jit/function.cpp:
```diff
@@ -46,19 +45,11 @@ IValue Function::operator()(
 }
 
 void Function::ensure_defined() {
```
There are call sites outside of the two callsites you've added error checking, is it valid to remove the checks there ? In `interpreter.cpp`, `schema_matching.cpp`, and `compiler.cpp`.

suo replied:
Yes, the only way we can call things recursively is either by calling a method or by calling a function. To be clear: a `RecursiveMethodCallError` exception will still be thrown if we managed to create recursion some other way, it'll just won't be printed with a nice hint and source location.

eellison replied:
true lol

eellison reviewed (commented):
Removing optional arg is a good change. i have a question about the error checking of recursion

suo committed:
Update on "[jit] Delete the ErrorReport default constructor"


An error report without a SourceRange context is bad, because it doesn't
tell the user where something happened. Delete the default constructor
to make it harder to create errors like this (you can still use a fake
SourceRange if you absolutely need to).

Also clean up the only case where the default constructor was used.

Differential Revision: [D19670924](https://our.internmc.facebook.com/intern/diff/D19670924)

[ghstack-poisoned]

suo committed:
Update on "[jit] Delete the ErrorReport default constructor"


An error report without a SourceRange context is bad, because it doesn't
tell the user where something happened. Delete the default constructor
to make it harder to create errors like this (you can still use a fake
SourceRange if you absolutely need to).

Also clean up the only case where the default constructor was used.

Differential Revision: [D19670924](https://our.internmc.facebook.com/intern/diff/D19670924)

[ghstack-poisoned]
